x/collection/msgs.go

```diff
@@ -0,0 +1,515 @@
+	}
+
+	return nil
+}
+
+func ValidateTokenID(id string) error {
+	if err := validateID(id, reTokenID); err != nil {
+		return ErrInvalidTokenID.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+func ValidateNFTID(id string) error {
+	if err := ValidateTokenID(id); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Deprecated: do not use (no successor).
+func ValidateLegacyNFTID(id string) error {
+	if err := validateID(id, reLegacyNFTID); err != nil {
+		return ErrInvalidTokenID.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+func ValidateLegacyIdxNFT(id string) error {
+	if err := validateID(id, reLegacyIdxNFTID); err != nil {
+		return ErrInvalidTokenID.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+func validateID(id string, reg *regexp.Regexp) error {
+	if !reg.MatchString(id) {
+		return sdkerrors.ErrInvalidRequest.Wrapf("invalid id: %s", id)
+	}
+	return nil
+}
+
+func validateName(name string) error {
+	if err := validateStringSize(name, nameLengthLimit, "name"); err != nil {
+		return ErrInvalidNameLength.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+func validateURI(uri string) error {
+	if err := validateStringSize(uri, uriLengthLimit, "uri"); err != nil {
+		return ErrInvalidBaseImgURILength.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+func validateMeta(meta string) error {
+	if err := validateStringSize(meta, metaLengthLimit, "meta"); err != nil {
+		return ErrInvalidMetaLength.Wrap(err.Error())
+	}
+
+	return nil
+}
+
+func validateStringSize(str string, limit int, name string) error {
+	if length := utf8.RuneCountInString(str); length > limit {
+		return sdkerrors.ErrInvalidRequest.Wrapf("%s cannot exceed %d in length: current %d", name, limit, length)
+	}
+	return nil
+}
+
+func validateLegacyPermission(permission string) error {
+	return ValidatePermission(Permission(LegacyPermissionFromString(permission)))
+}
+
+func ValidatePermission(permission Permission) error {
+	if p := Permission_value[Permission_name[int32(permission)]]; p == 0 {
+		return ErrInvalidPermission.Wrap(permission.String())
+	}
+	return nil
+}
+
+func validateContractChange(change Attribute) error {
+	validators := map[string]func(string) error{
+		AttributeKeyName.String():       validateName,
+		AttributeKeyBaseImgURI.String(): validateURI,
+		AttributeKeyMeta.String():       validateMeta,
+		AttributeKeyURI.String():        validateURI,
+	}
+
+	return validateChange(change, validators)
+}
+
+func validateTokenClassChange(change Attribute) error {
+	validators := map[string]func(string) error{
+		AttributeKeyName.String(): validateName,
+		AttributeKeyMeta.String(): validateMeta,
+	}
+
+	return validateChange(change, validators)
+}
+
+func validateChange(change Attribute, validators map[string]func(string) error) error {
+	validator, ok := validators[change.Key]
+	if !ok {
+		return ErrInvalidChangesField.Wrapf("invalid field: %s", change.Key)
+	}
+	return validator(change.Value)
+}
+
+// ValidateBasic implements Msg.
+func (m MsgSendNFT) ValidateBasic() error {
```

Comment:
After applying address codec to this module, we must update the signature to accept it.